src/librustc_resolve/diagnostics.rs

```diff
@@ -857,54 +858,173 @@ impl<'a> Resolver<'a> {
+
+                suggestions.extend(self.lookup_import_candidates_from_module(
+                    lookup_ident,
+                    namespace,
+                    parent_scope,
+                    crate_root,
+                    ident,
+                    &filter_fn,
+                    &mut collect_fn,
+                ));
             }
         }
 
         suggestions
     }
 
+    /// The purpose of this function is to collect public symbols names that are unique across all
+    /// crates in the build. Later, when printing about types we can use those names instead of the
+    /// full exported path to them.
+    ///
+    /// So essentially, if a symbol name can only be imported from one place for a type, and as
+    /// long as it was not glob-imported anywhere in the current crate, we can trim its printed
+    /// path and print only the name.
+    ///
+    /// This has wide implications on error messages with types, for example, shortening
+    /// `std::vec::Vec` to just `Vec`, as long as there is no other `Vec` importable anywhere.
+    ///
+    /// The implementation uses that same underyling mechanism that 'use' suggestions are using,
+    /// and there a few aspects to how this is done:
+    ///
+    /// * Ignore type aliases. This allows `std::result::Result` to compete over the
+    ///   `std::io::Result` alias.
+    /// * Separate uniqueness for trait symbols and type symbols. This sorts out an overlap between
+    ///   `std::fmt::Display` and `std::path::Display`, for instance.
+    /// * Skip Variant types. Othersize, it can pick up `core::num::flt2dev::Part::Copy` as collsion
+    ///   with `std::marker::Copy`. It also makes sense, as users cannot specifically import this
+    ///   into the Type namespace (rather than the Value namespace).
+    /// * Disambiguate module names having the same name as functions.
+    pub fn find_unique_symbols(&mut self) {
+        if self.session.opts.debugging_opts.disable_unique_symbols {
+            return;
+        }
+
+        self.avoid_crate_load = true;
+
+        #[derive(Hash, Eq, PartialEq, Copy, Clone)]
+        enum Kind {
+            Value,
+            Type,
+            Trait,
+            Macro,
+        };
+
+        let graph_root = self.graph_root;
+        let parent_scope = ParentScope::module(graph_root);
+        let unique_symbols: &mut FxHashMap<(Kind, Symbol), Option<DefId>> =
+            &mut FxHashMap::default();
+
+        for symbol in self.glob_imported.drain() {
+            unique_symbols.insert((Kind::Value, symbol), None);
+            unique_symbols.insert((Kind::Type, symbol), None);
+            unique_symbols.insert((Kind::Trait, symbol), None);
+            unique_symbols.insert((Kind::Macro, symbol), None);
+        }
+
+        self.lookup_import_candidates(
+            Ident::invalid(),
+            TypeNS,
+            &parent_scope,
+            |_| false,
```

Comment:
I'm not sure reusing `lookup_import_candidates` is a good idea here.
All its arguments end up being unused, basically, except for `parent_scope` which ends up being used incorrectly.

Could you keep `lookup_import_candidates` alone, and introduce a new function that visits items in all crates like `lookup_import_candidates`, but drops all the activities not necessary for `find_unique_symbols`?
(We may need to do that for performance anyway.)

Reply:
(Then `avoid_crate_load` will become entirely unnecessary as well.)

Reply:
Done. The new function's implementation is shorter and may be faster.